trace/span.proto

```diff
@@ -0,0 +1,323 @@
+    //
+    //     "/instance_id": "my-instance"
+    //     "/http/user_agent": ""
+    //     "/http/server_latency": 300
+    //     "abc.com/myattribute": true
+    map<string, AttributeValue> attribute_map = 1;
+
+    // The number of attributes that were discarded. Attributes can be discarded
+    // because their keys are too long or because there are too many attributes.
+    // If this value is 0 then all attributes are valid.
+    int32 dropped_attributes_count = 2;
+  }
+
+  // A set of attributes on the span.
+  Attributes attributes = 7;
+
+  // Stack trace captured at the start of the span.
+  StackTrace stack_trace = 8;
+
+  // A time-stamped annotation or network event in the Span.
+  message TimeEvent {
+    // The timestamp indicating the time the event occurred.
+    google.protobuf.Timestamp time = 1;
+
+    // Text annotation with a set of attributes.
+    message Annotation {
+      // A user-supplied message describing the event.
+      TruncatableString description = 1;
+
+      // A set of attributes on the annotation.
+      Attributes attributes = 2;
+    }
+
+    // An event describing an RPC message sent or received on the network.
+    message NetworkEvent {
+      // For sent messages, this is the time at which the first bit was sent.
+      // For received messages, this is the time at which the last bit was
+      // received.
+      google.protobuf.Timestamp time = 1;
+
+      // Indicates whether the network message was sent or received.
+      enum Type {
+        // Unknown event type.
+        TYPE_UNSPECIFIED = 0;
+        // Indicates a sent RPC message.
+        SENT = 1;
+        // Indicates a received RPC message.
+        RECV = 2;
+      }
+
+      // Type of NetworkEvent. Indicates whether the RPC message was sent or
+      // received.
+      Type type = 2;
+
+      // An identifier for the message, which must be unique in this span.
+      uint64 message_id = 3;
+
+      // The number of uncompressed bytes sent or received.
+      uint64 uncompressed_message_size = 4;
+
+      // The number of compressed bytes sent or received.
+      uint64 compressed_message_size = 5;
+    }
+
+    // A `TimeEvent` can contain either an `Annotation` object or a
+    // `NetworkEvent` object, but not both.
+    oneof value {
+      // One or more key:value pairs.
+      Annotation annotation = 2;
+
+      // An event describing an RPC message sent/received on the network.
+      NetworkEvent network_event = 3;
+    }
+  }
+
+  // A collection of `TimeEvent`s. A `TimeEvent` is a time-stamped annotation
+  // on the span, consisting of either user-supplied key:value pairs, or
+  // details of an RPC message sent/received on the network.
+  message TimeEvents {
+    // A collection of `TimeEvent`s.
+    repeated TimeEvent time_event = 1;
+
+    // The number of dropped annotations in all the included time events.
+    // If the value is 0, then no annotations were dropped.
+    int32 dropped_annotations_count = 2;
```

Comment:
this looks maintenance causing, which might be ok. For example, if another type of event is added, you need a new field.

Reply:
unfortunately proto does not have generics 👎